coding_conventions.md

```diff
@@ -0,0 +1,154 @@
+# Coding conventions
+
+The following is a list of conventions we follow when working in the [water-abstraction-system](https://github.com/DEFRA/water-abstraction-system) repo.
+
+We make best efforts to follow them when working in the legacy repos. There are often times it makes more sense to follow the conventions that already exist.
+
+- [Add the .js extension](#add-the-js-extension)
+- [File names](#file-names)
+- [Functions](#functions)
+- [Top of .js files](#top-of-js-files)
+  - [Top of test.js files](#top-of-testjs-files)
+
+## Add the .js extension
+
+When linking to an internal module in our `require()` statements _always_ include the `*.js` extension.
+
+```javascript
+// Good
+const RequestLib = require('../lib/request.lib.js')
+
+// Bad
+const RequestLib = require('../lib/request.lib')
+```
+
+We _know_ 😝 that you don't have to and that generally the convention is not to. But we still hold out hope that one day we'll [switch our main repo from CommonJS to ES6 modules](https://github.com/DEFRA/water-abstraction-system). The extension _is_ required when using `import` and the work of migrating is made easier if the existing `require()` statements already include it.
+
+## File names
+
+All file names should be [kebab-case](https://www.freecodecamp.org/news/snake-case-vs-camel-case-vs-pascal-case-vs-kebab-case-whats-the-difference/#kebab-case), for example `awesome-people.service.js`
+
+Other naming conventions are
+
+- controllers `thing.controller.js`
+- general modules `thing.lib.js`
+- models `thing.model.js`
+- plugins `thing.plugin.js`
+- presenters `thing.presenter.js`
+- routes `thing.routes.js`
+- services `thing.service.js`
+
+Unit test files should be the same as the thing being tested with `.test` added to the end, for example `thing.service.test.js`. The `test/` folder structure should mirror the `app/` folder.
+
+## Functions
+
+We define our functions as
+
+```javascript
+function helloWorld () {
+  return 'Hello, world!'
+}
+
+// DON'T do this
+const helloWorld = () => {
+  return 'Hello, world!'
+}
+```
+
+When we do need to use arrow functions, parameters are always wrapped in brackets even if there is only one. Also, use _block body_ over _concise body_.
+
+```javascript
+const materials = [
+  'Hydrogen',
+  'Helium',
+  'Lithium',
+  'Beryllium'
+]
+
+// Use Block body version
+materials.forEach((material) => {
+  console.log(`${material} - ${material.length}`)
+})
+
+// Do not use Concise body version
+materials.forEach((material) => console.log(`${material} - ${material.length}`))
```

Comment:
I'd forgotten about this rule and may have given Rebecca a bum steer when going through some Lodash code yesterday! But I just wanted to check if we always always want to avoid concise body version. Just because personally I prefer chained methods looking like this:
```js
  const concise = iterable
    .map((item) => parseFloat(item[qtyKey]))
    .filter((value) => ![null, undefined, NaN].includes(value))
    .reduce((acc, num) => acc + num, 0)
```

Over chained methods looking like this:
```js
  const block = iterable
  .map((item) => {
    return parseFloat(item[qtyKey])
  })
  .filter((value) => {
    return ![null, undefined, NaN].includes(value)
  })
  .reduce((acc, num) => {
    return acc + num
  }, 0)
```

But this is personal preference and I can live with the latter if needs be.

Reply:
I'm going to stick with the convention (don't use concise body). I've got 2 reasons (whether they are good or not is up for debate!)

First, you would be in the majority of developers who would prefer to see the concise version. But most developers also tend to work on static 'products', or at least in static teams. When it comes to us, we shift around both teams and products frequently. Add to that we often don't make much distinction between the experience and skills of our developers, throwing whoever is available at whatever legacy project needs work.

So, I _try_ and approach things as if I'm a junior dev, new to a project, with only a few `Hello, world!` tutorials in the chosen language under my belt. Using this example, you know something is being returned because it says so. You don't have to know about single-line variant to figure out what's happening.

The second is 'code smell'. I use shape and length as primary indicators as to whether something is too complex or doing too much. You'd be forgiven for overlooking the concise version in the middle of a method. But the expanded version would be a stronger 'smell' that it needs refactoring to its own method.

So, I know our convention wouldn't be the community convention. But I'd argue an experienced JavaScript developer wouldn't be impeded by this (though they'd probably 🤮 !) whereas a junior _might_ be helped.

Reply:
See what you did @StuAA78 !? This has been around so long @Jozzey went and found an exception to the rule in [Determine if licence is 'billed'](https://github.com/DEFRA/water-abstraction-system/pull/82) 😩 🤦 

So, where we're just doing a 'truthy' check I think it is clearer when written concise-body style

```javascript
const alansSkills = [
  { name: 'Waffle', level: 'High' },
  { name: 'BS' }
]
const stuartsSkills = [
  { name: 'Waffle' },
  { name: 'BS' }
]

// Concise body - IMHO it's clear we only care if it's 'truthy'
const alanRatingsComplete = alansSkills.some((skill) => skill.level)

// Block body - IMHO this makes it look like we care more about the value of `skill.level`
const stuartHasSkillLevels = stuartsSkills.some((skill) => {
  return skill.level
})

console.log(`Alan's skills are all rated: ${alanHasSkillLevels}`)
console.log(`Stuart's skills are all rated: ${stuartHasSkillLevels}`)
```

Assuming you agree, I'll add this exception but I really want to get this merged. So, I'll add it to the list of amendments and additions that need making.

What do you think?